src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -1153,30 +1152,59 @@ private static ulong TransferDataFromRequestStream(IntPtr buffer, int length, Ea
                 Debug.Assert(sts._offset == 0, "Offset should be zero.");
 
                 // If we get here, there was no previously read data available to copy.
-                // Initiate a new asynchronous read.
-                Task<int> asyncRead = easy._requestContentStream.ReadAsyncInternal(
-                    sts._buffer, 0, Math.Min(sts._buffer.Length, length), easy._cancellationToken);
+
+                // Make sure we actually have a stream to read from.  This will be null if either
+                // this is the first time we're reading it, or if the stream was reset as part
+                // of curl trying to rewind.  Then do the read.
+                Task<int> asyncRead;
+                if (easy._requestContentStream == null)
+                {
+                    multi.EventSourceTrace("Calling ReadAsStreamAsync to get new request stream", easy: easy);
+                    asyncRead = easy._requestMessage.Content.ReadAsStreamAsync().ContinueWith((readStream, s) =>
+                    {
+                        multi.EventSourceTrace("ReadAsStreamAsync completed: {0}", readStream.Status, easy: easy);
+
+                        var stateAndRequest = (Tuple<EasyRequest.SendTransferState, EasyRequest>)s;
+                        stateAndRequest.Item2._requestContentStream = readStream.GetAwaiter().GetResult();
+                        multi.EventSourceTrace("Got stream: {0}", stateAndRequest.Item2._requestContentStream.GetType(), easy: easy);
+
+                        multi.EventSourceTrace("Starting async read", easy: easy);
+                        return stateAndRequest.Item2._requestContentStream.ReadAsync(
+                           stateAndRequest.Item1._buffer, 0, Math.Min(stateAndRequest.Item1._buffer.Length, length), stateAndRequest.Item2._cancellationToken);
+                    }, Tuple.Create(sts, easy), easy._cancellationToken, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.DenyChildAttach, TaskScheduler.Default).Unwrap();
```

Comment:
A) :smile:
B) Thanks.  If I have other reasons to resubmit it, I'll reformat the line.